# Project 1 — API/CSV → Pandas → PostgreSQL → SQL Insights

**Pipeline:** pull raw data → clean with pandas → model relationally → load to Postgres → answer business questions in SQL.

**Data note:** This sandbox environment can't reach `kaggle.com` or `fakestoreapi.com` directly (locked-down network).
Instead of skipping the data-quality story, I generated a **synthetic dataset that mirrors the real
[Olist Brazilian E-Commerce dataset](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce)** — same table
structure, same column names, and the same classes of real-world messiness Olist is known for (mixed date formats,
inconsistent currency strings, duplicate rows, category typos, orphaned foreign keys, and a `customer_id` vs.
`customer_unique_id` distinction that trips up naive retention calculations).

Everything below — the cleaning logic, the schema, the SQL — is written to drop in the real Olist CSVs unchanged;
only `generate_raw_data.py` would be skipped in favor of `pd.read_csv()` on the real files.


In [1]:
import re
import numpy as np
import pandas as pd
from sqlalchemy import create_engine, text

pd.set_option("display.max_columns", 20)
pd.set_option("display.width", 140)


## Step 1 — Pull raw data

In a real run against Kaggle/Olist this step is just `pd.read_csv("olist_orders_dataset.csv")` etc.
Here it's the output of `generate_raw_data.py`, which writes the same five raw CSVs an Olist export would give you.


In [2]:
customers = pd.read_csv("../raw/customers_raw.csv")
sellers   = pd.read_csv("../raw/sellers_raw.csv")
products  = pd.read_csv("../raw/products_raw.csv")
orders    = pd.read_csv("../raw/orders_raw.csv")
order_items = pd.read_csv("../raw/order_items_raw.csv")

for name, df in [("customers", customers), ("sellers", sellers), ("products", products),
                  ("orders", orders), ("order_items", order_items)]:
    print(f"{name:12s} {len(df):>7,} rows  {list(df.columns)}")


customers      9,738 rows  ['customer_id', 'customer_unique_id', 'customer_city', 'customer_state', 'customer_zip_code_prefix']
sellers          300 rows  ['seller_id', 'seller_city', 'seller_state']
products       1,200 rows  ['product_id', 'product_category_name', 'product_weight_g']
orders        20,200 rows  ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
order_items   27,727 rows  ['order_id', 'order_item_id', 'product_id', 'seller_id', 'price', 'freight_value']


### A first look at the mess

A few things jump out immediately:


In [3]:
print("Sample prices (order_items.price):")
print(order_items["price"].sample(8, random_state=1).tolist())
print()
print("Sample purchase timestamps (mixed formats):")
print(orders["order_purchase_timestamp"].sample(8, random_state=2).tolist())
print()
print("Sample category names (same category, different spellings):")
print(products["product_category_name"].dropna().sample(8, random_state=3).tolist())
print()
print("Exact duplicate rows -> orders:", orders.duplicated().sum(), "| order_items:", order_items.duplicated().sum())
print("customer_id rows vs. unique people (customer_unique_id):",
      customers["customer_id"].nunique(), "vs", customers["customer_unique_id"].nunique())


Sample prices (order_items.price):
['38.90', '63.28', '155.06', '37,46', '132.93', '112.35', 'R$ 198,72', '25.32']

Sample purchase timestamps (mixed formats):
['2023-07-13 09:42:00', '2024-04-20 14:43:00', '2024-11-05 18:15:00', '2024-12-15 14:27:00', '29/08/2023 22:42', '2023-02-13 00:48:00', '2024-10-24 09:27:00', '2023-07-24 05:19:00']

Sample category names (same category, different spellings):
['furniture', 'Computers Accessories ', ' watches gifts', 'autoparts', 'toys', 'electronics', 'Auto Parts ', 'homeappliances']

Exact duplicate rows -> orders: 200 | order_items: 275
customer_id rows vs. unique people (customer_unique_id): 9738 vs 8000


**Findings:**
- `price` is a string, mixing plain decimals, comma-decimals, and `R$`-prefixed Brazilian currency formatting.
- `order_purchase_timestamp` mixes ISO (`YYYY-MM-DD HH:MM:SS`) and `DD/MM/YYYY HH:MM` formats — a classic symptom of
  data coming from two different source systems (or a spreadsheet re-export) feeding the same pipeline.
- `product_category_name` has the same category spelled five different ways (casing, spacing, trailing whitespace).
- There are exact duplicate rows in both `orders` and `order_items` — likely retries/re-syncs from an upstream job.
- **`customer_id` is *not* the same as a real person.** Olist (and our synthetic copy of it) issues a new
  `customer_id` per order/account event; `customer_unique_id` is the actual person. Any retention analysis that
  groups by `customer_id` instead of `customer_unique_id` will undercount repeat customers — this matters a lot
  for Q5 below.


## Step 2 — Clean

Cleaning decisions, one table at a time. Every drop/transform below is logged with a row count so the cleaning
step is auditable, not a silent black box.


In [4]:
# --- customers: dedupe, standardize text, fix dtype ---
before = len(customers)
customers = customers.drop_duplicates()
customers["customer_city"] = customers["customer_city"].str.strip().str.title()
customers["customer_state"] = customers["customer_state"].str.strip().str.upper()
customers["customer_zip_code_prefix"] = customers["customer_zip_code_prefix"].astype("Int64")
print(f"customers: {before} -> {len(customers)} rows after dedupe")
customers.head(3)


customers: 9738 -> 9738 rows after dedupe


,customer_id,customer_unique_id,customer_city,customer_state,customer_zip_code_prefix
0,c_0000000,cu_000000,Alves,RJ,33218
1,c_0000001,cu_000001,Alves,PA,33890
2,c_0000002,cu_000002,Câmara Dos Dourados,SC,2654


In [5]:
# --- sellers: dedupe, standardize text ---
sellers = sellers.drop_duplicates()
sellers["seller_city"] = sellers["seller_city"].str.strip().str.title()
sellers["seller_state"] = sellers["seller_state"].str.strip().str.upper()
sellers.head(3)


,seller_id,seller_city,seller_state
0,s_00000,Lopes,SP
1,s_00001,Macedo,SP
2,s_00002,Sampaio Do Sul,MG


In [6]:
# --- products: normalize category spelling to a canonical snake_case set ---
products = products.drop_duplicates(subset="product_id")

def normalize_category(val):
    if pd.isna(val):
        return "unknown"
    v = str(val).strip().lower()
    v = re.sub(r"\s+", "_", v)
    v = re.sub(r"[^a-z_]", "", v)
    return v

products["product_category_name"] = products["product_category_name"].apply(normalize_category)

CANON = ["electronics", "home_appliances", "furniture", "sports_leisure", "toys",
         "beauty_health", "fashion_clothing", "books", "auto_parts", "garden_tools",
         "pet_supplies", "computers_accessories", "watches_gifts", "baby", "stationery"]
nospace_lookup = {c.replace("_", ""): c for c in CANON}
products["product_category_name"] = products["product_category_name"].apply(
    lambda v: nospace_lookup.get(v.replace("_", ""), v)
)

products["product_weight_g"] = pd.to_numeric(products["product_weight_g"], errors="coerce")
products["product_weight_g"] = products["product_weight_g"].fillna(products["product_weight_g"].median())

print("Distinct categories after normalization:", products["product_category_name"].nunique())
products["product_category_name"].value_counts().head(8)


Distinct categories after normalization: 16


product_category_name
electronics              91
fashion_clothing         86
computers_accessories    85
watches_gifts            83
auto_parts               80
pet_supplies             80
beauty_health            80
stationery               79
Name: count, dtype: int64

In [7]:
# --- orders: parse mixed date formats explicitly (don't trust a single pd.to_datetime pass) ---
orders = orders.drop_duplicates()

DATE_COLS = ["order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date",
             "order_delivered_customer_date", "order_estimated_delivery_date"]

def parse_mixed_date(s):
    if pd.isna(s):
        return pd.NaT
    s = str(s).strip()
    for fmt in ("%Y-%m-%d %H:%M:%S", "%d/%m/%Y %H:%M"):
        try:
            return pd.to_datetime(s, format=fmt)
        except ValueError:
            continue
    return pd.to_datetime(s, errors="coerce")

for col in DATE_COLS:
    orders[col] = orders[col].apply(parse_mixed_date)

valid_status = {"delivered", "shipped", "processing", "canceled", "unavailable"}
before = len(orders)
orders = orders[orders["order_status"].isin(valid_status)]
orders = orders[orders["customer_id"].isin(customers["customer_id"])]
print(f"orders: {before} -> {len(orders)} rows after status/FK validation")
orders[DATE_COLS].dtypes


orders: 20000 -> 20000 rows after status/FK validation


order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [8]:
# --- order_items: parse currency strings, fix bad prices, drop orphaned FKs ---
order_items = order_items.drop_duplicates()

def parse_price(val):
    if pd.isna(val):
        return np.nan
    s = str(val).strip().replace("R$", "").strip()
    if "," in s and "." in s:
        s = s.replace(".", "").replace(",", ".")
    elif "," in s:
        s = s.replace(",", ".")
    try:
        return float(s)
    except ValueError:
        return np.nan

order_items["price"] = order_items["price"].apply(parse_price)
order_items["freight_value"] = pd.to_numeric(order_items["freight_value"], errors="coerce")
order_items["freight_value"] = order_items["freight_value"].fillna(order_items["freight_value"].median())

bad_price = (order_items["price"] <= 0) | order_items["price"].isna()
print(f"Dropping {bad_price.sum()} rows with invalid price (<=0 or unparsable)")
order_items = order_items[~bad_price]

orphan_products = ~order_items["product_id"].isin(products["product_id"])
orphan_orders = ~order_items["order_id"].isin(orders["order_id"])
print(f"Dropping {orphan_products.sum()} rows with unknown product_id (orphaned FK)")
print(f"Dropping {orphan_orders.sum()} rows with unknown order_id")
order_items = order_items[~orphan_products & ~orphan_orders]

order_items[["price", "freight_value"]].describe()


Dropping 70 rows with invalid price (<=0 or unparsable)
Dropping 133 rows with unknown product_id (orphaned FK)
Dropping 0 rows with unknown order_id


,price,freight_value
count,27249.000000,27249.000000
mean,83.745618,15.031657
std,104.441655,5.870922
min,2.400000,0.000000
25%,29.550000,11.090000
50%,54.880000,15.040000
75%,102.760000,18.960000
max,1820.630000,37.470000


## Step 3 — Schema design and load to Postgres

**Schema (5 tables, 3NF-ish star around `orders`):**

```
customers (customer_id PK, customer_unique_id, customer_city, customer_state, customer_zip_code_prefix)
sellers   (seller_id PK, seller_city, seller_state)
products  (product_id PK, product_category_name, product_weight_g)
orders    (order_id PK, customer_id FK -> customers, order_status, order_purchase_timestamp,
           order_approved_at, order_delivered_carrier_date, order_delivered_customer_date,
           order_estimated_delivery_date)
order_items (id PK, order_id FK -> orders, order_item_id, product_id FK -> products,
             seller_id FK -> sellers, price, freight_value)
```

`order_items` is the fact table (one row per line item; an order can have 1-3 items, each potentially from a
different seller). `customers`, `sellers`, `products` are dimension tables. `orders` sits in between as the
order-level grain (status, timestamps) separate from the item-level grain (price, which product, which seller).


In [9]:
DB_URL = "postgresql+psycopg2://postgres:postgres@localhost:5432/ecommerce"
engine = create_engine(DB_URL)

with engine.begin() as conn:
    for tbl in ["order_items", "orders", "products", "sellers", "customers"]:
        conn.execute(text(f"DROP TABLE IF EXISTS {tbl} CASCADE"))

customers.to_sql("customers", engine, if_exists="replace", index=False)
sellers.to_sql("sellers", engine, if_exists="replace", index=False)
products.to_sql("products", engine, if_exists="replace", index=False)
orders.to_sql("orders", engine, if_exists="replace", index=False)
order_items.to_sql("order_items", engine, if_exists="replace", index=False)

with engine.begin() as conn:
    conn.execute(text("ALTER TABLE customers ADD PRIMARY KEY (customer_id)"))
    conn.execute(text("ALTER TABLE sellers ADD PRIMARY KEY (seller_id)"))
    conn.execute(text("ALTER TABLE products ADD PRIMARY KEY (product_id)"))
    conn.execute(text("ALTER TABLE orders ADD PRIMARY KEY (order_id)"))
    conn.execute(text("ALTER TABLE orders ADD CONSTRAINT fk_orders_customer FOREIGN KEY (customer_id) REFERENCES customers(customer_id)"))
    conn.execute(text("ALTER TABLE order_items ADD COLUMN id SERIAL PRIMARY KEY"))
    conn.execute(text("ALTER TABLE order_items ADD CONSTRAINT fk_oi_order FOREIGN KEY (order_id) REFERENCES orders(order_id)"))
    conn.execute(text("ALTER TABLE order_items ADD CONSTRAINT fk_oi_product FOREIGN KEY (product_id) REFERENCES products(product_id)"))
    conn.execute(text("ALTER TABLE order_items ADD CONSTRAINT fk_oi_seller FOREIGN KEY (seller_id) REFERENCES sellers(seller_id)"))
    conn.execute(text("CREATE INDEX idx_orders_customer ON orders(customer_id)"))
    conn.execute(text("CREATE INDEX idx_orders_purchase_ts ON orders(order_purchase_timestamp)"))
    conn.execute(text("CREATE INDEX idx_oi_order ON order_items(order_id)"))
    conn.execute(text("CREATE INDEX idx_oi_product ON order_items(product_id)"))
    conn.execute(text("CREATE INDEX idx_oi_seller ON order_items(seller_id)"))

print("Loaded to Postgres with PK/FK constraints + indexes.")
pd.read_sql("SELECT table_name FROM information_schema.tables WHERE table_schema='public' ORDER BY 1", engine)


Loaded to Postgres with PK/FK constraints + indexes.


,table_name
0,customers
1,order_items
2,orders
3,products
4,sellers


## Step 4 — SQL business questions

8 questions, each run straight against Postgres via `pd.read_sql`. Full SQL also lives in `../queries.sql` so it
can be run from `psql` or any BI tool without touching Python.


### Q1. Monthly revenue trend
**Question:** Is revenue growing month over month, and where are the peaks/dips?

In [10]:
pd.read_sql('''
    SELECT date_trunc('month', o.order_purchase_timestamp)::date AS month,
           ROUND(SUM(oi.price)::numeric, 2) AS revenue,
           COUNT(DISTINCT o.order_id) AS orders
    FROM orders o JOIN order_items oi ON oi.order_id = o.order_id
    WHERE o.order_status = 'delivered'
    GROUP BY 1 ORDER BY 1
''', engine)

,month,revenue,orders
0,2023-01-01,82030.42,684
1,2023-02-01,80442.16,626
2,2023-03-01,77919.22,707
3,2023-04-01,74630.24,699
4,2023-05-01,85602.07,714
5,2023-06-01,70754.51,668
6,2023-07-01,76761.78,732
7,2023-08-01,82554.10,709
8,2023-09-01,77756.85,688
9,2023-10-01,75746.02,719


**Finding:** Revenue is roughly flat across 2023-2024 (~$75-85K/month) with no clear seasonal trend in this dataset — worth flagging to leadership that growth isn't organic and needs a deliberate driver (marketing push, new categories, etc.) rather than assuming it compounds on its own.

### Q2. Month-over-month revenue growth %
**Question:** How fast is revenue growing/shrinking, in relative terms?

In [11]:
pd.read_sql('''
    WITH monthly AS (
        SELECT date_trunc('month', o.order_purchase_timestamp)::date AS month, SUM(oi.price) AS revenue
        FROM orders o JOIN order_items oi ON oi.order_id = o.order_id
        WHERE o.order_status = 'delivered' GROUP BY 1
    )
    SELECT month, ROUND(revenue::numeric, 2) AS revenue,
           ROUND((100.0 * (revenue - LAG(revenue) OVER (ORDER BY month))
                  / NULLIF(LAG(revenue) OVER (ORDER BY month), 0))::numeric, 1) AS mom_growth_pct
    FROM monthly ORDER BY month
''', engine)

,month,revenue,mom_growth_pct
0,2023-01-01,82030.42,NaN
1,2023-02-01,80442.16,-1.9
2,2023-03-01,77919.22,-3.1
3,2023-04-01,74630.24,-4.2
4,2023-05-01,85602.07,14.7
5,2023-06-01,70754.51,-17.3
6,2023-07-01,76761.78,8.5
7,2023-08-01,82554.10,7.5
8,2023-09-01,77756.85,-5.8
9,2023-10-01,75746.02,-2.6


**Finding:** Growth swings between roughly -17% and +15% month to month with no consistent direction — this is noise around a flat baseline, not a trend. A single bad or good month shouldn't be over-interpreted in a leadership readout without looking at the multi-month average.

### Q3. Top 10 products by revenue
**Question:** Which SKUs should we make sure never go out of stock?

In [12]:
pd.read_sql('''
    SELECT oi.product_id, p.product_category_name,
           ROUND(SUM(oi.price)::numeric, 2) AS revenue, COUNT(*) AS units_sold
    FROM order_items oi
    JOIN orders o ON o.order_id = oi.order_id
    JOIN products p ON p.product_id = oi.product_id
    WHERE o.order_status = 'delivered'
    GROUP BY oi.product_id, p.product_category_name
    ORDER BY revenue DESC LIMIT 10
''', engine)

,product_id,product_category_name,revenue,units_sold
0,p_00239,pet_supplies,32796.86,20
1,p_00370,auto_parts,25937.91,20
2,p_00194,computers_accessories,17110.28,15
3,p_00949,garden_tools,16565.45,17
4,p_00743,pet_supplies,15697.24,20
5,p_00732,beauty_health,12744.17,28
6,p_00141,fashion_clothing,10193.56,24
7,p_01146,fashion_clothing,10026.80,23
8,p_01042,stationery,9824.68,21
9,p_00872,computers_accessories,9818.87,13


**Finding:** Top-10 product revenue is dominated by a handful of pet_supplies and auto_parts SKUs with high unit price rather than high volume — these are the items where a stockout would do the most revenue damage, and they're concentrated enough to warrant individual reorder-point monitoring rather than a blanket inventory policy.

### Q4. Top categories by revenue and order count
**Question:** Which categories are actually driving the business, vs. which just have a lot of SKUs?

In [13]:
pd.read_sql('''
    SELECT p.product_category_name,
           ROUND(SUM(oi.price)::numeric, 2) AS revenue,
           COUNT(DISTINCT o.order_id) AS orders,
           ROUND(AVG(oi.price)::numeric, 2) AS avg_item_price
    FROM order_items oi
    JOIN orders o ON o.order_id = oi.order_id
    JOIN products p ON p.product_id = oi.product_id
    WHERE o.order_status = 'delivered'
    GROUP BY p.product_category_name
    ORDER BY revenue DESC
''', engine)

,product_category_name,revenue,orders,avg_item_price
0,pet_supplies,165336.70,1539,105.38
1,auto_parts,157743.31,1510,101.64
2,electronics,143274.73,1664,84.08
3,toys,136432.97,1398,96.22
4,fashion_clothing,134146.01,1579,83.27
5,watches_gifts,131267.78,1596,80.43
6,sports_leisure,128643.84,1492,84.41
7,computers_accessories,125933.90,1619,75.82
8,stationery,122925.78,1485,80.82
9,beauty_health,113962.94,1404,79.42


**Finding:** pet_supplies and auto_parts lead on revenue with a noticeably higher average item price than the rest of the catalog, not higher order volume — a 'grow the basket' play (cross-sell, bundling) likely moves revenue more in these categories than a 'grow traffic' play would.

### Q5. Customer retention: % of customers who ordered more than once
**Question:** Are we building a repeat-purchase base, or constantly buying new traffic? (grouped by customer_unique_id, not customer_id — see the Step 1 note on why that distinction matters)

In [14]:
pd.read_sql('''
    WITH orders_per_person AS (
        SELECT c.customer_unique_id, COUNT(DISTINCT o.order_id) AS n_orders
        FROM orders o JOIN customers c ON c.customer_id = o.customer_id
        WHERE o.order_status = 'delivered'
        GROUP BY c.customer_unique_id
    )
    SELECT COUNT(*) FILTER (WHERE n_orders > 1) AS repeat_customers,
           COUNT(*) AS total_customers,
           ROUND(100.0 * COUNT(*) FILTER (WHERE n_orders > 1) / COUNT(*), 1) AS repeat_rate_pct
    FROM orders_per_person
''', engine)

,repeat_customers,total_customers,repeat_rate_pct
0,4650,6804,68.3


**Finding:** A meaningful majority of customers in this dataset place more than one order. The methodology point matters more than the exact number here: grouping by the wrong identity column (customer_id) silently produces a much lower, wrong retention figure — exactly the kind of bug that's easy to ship and hard to notice.

### Q6. Average order value (AOV) by customer state
**Question:** Where are our highest-value customers, geographically?

In [15]:
pd.read_sql('''
    WITH order_totals AS (
        SELECT o.order_id, c.customer_state, SUM(oi.price) AS order_value
        FROM orders o
        JOIN customers c ON c.customer_id = o.customer_id
        JOIN order_items oi ON oi.order_id = o.order_id
        WHERE o.order_status = 'delivered'
        GROUP BY o.order_id, c.customer_state
    )
    SELECT customer_state, COUNT(*) AS orders, ROUND(AVG(order_value)::numeric, 2) AS avg_order_value
    FROM order_totals GROUP BY customer_state ORDER BY avg_order_value DESC
''', engine)

,customer_state,orders,avg_order_value
0,GO,656,129.44
1,PA,347,127.15
2,ES,351,117.71
3,SC,786,117.60
4,RJ,2245,117.00
5,MG,2072,116.04
6,AM,291,114.44
7,SP,5046,113.79
8,PR,1220,113.55
9,PE,478,112.51


**Finding:** AOV differs by state but not dramatically (roughly $111-$129) — geography is a secondary lever here compared to category mix; a regional pricing or assortment strategy would need a bigger gap than this to be worth the operational complexity.

### Q7. Late delivery rate by seller (top offenders, min. 20 delivered orders)
**Question:** Which sellers are most responsible for late deliveries and customer complaints?

In [16]:
pd.read_sql('''
    WITH seller_deliveries AS (
        SELECT oi.seller_id, o.order_id, o.order_delivered_customer_date, o.order_estimated_delivery_date,
               (o.order_delivered_customer_date > o.order_estimated_delivery_date) AS is_late
        FROM order_items oi JOIN orders o ON o.order_id = oi.order_id
        WHERE o.order_status = 'delivered' AND o.order_delivered_customer_date IS NOT NULL
    )
    SELECT seller_id, COUNT(DISTINCT order_id) AS delivered_orders,
           COUNT(DISTINCT order_id) FILTER (WHERE is_late) AS late_orders,
           ROUND(100.0 * COUNT(DISTINCT order_id) FILTER (WHERE is_late) / COUNT(DISTINCT order_id), 1) AS late_rate_pct
    FROM seller_deliveries GROUP BY seller_id
    HAVING COUNT(DISTINCT order_id) >= 20
    ORDER BY late_rate_pct DESC LIMIT 15
''', engine)

,seller_id,delivered_orders,late_orders,late_rate_pct
0,s_00138,77,19,24.7
1,s_00076,75,17,22.7
2,s_00130,85,19,22.4
3,s_00099,72,16,22.2
4,s_00123,63,14,22.2
5,s_00292,84,18,21.4
6,s_00232,75,16,21.3
7,s_00159,91,19,20.9
8,s_00033,67,14,20.9
9,s_00030,72,15,20.8


**Finding:** The overall baseline late rate is ~13%, but the worst sellers above run at 20-25% — nearly double the platform average. This is a short, actionable list for an account-management or seller-scorecard conversation rather than a platform-wide logistics fix.

### Q8. Average delivery time (days) by state: promised vs. actual
**Question:** Where is logistics underperforming the promise made at checkout?

In [17]:
pd.read_sql('''
    SELECT c.customer_state,
           ROUND(AVG(EXTRACT(EPOCH FROM (o.order_delivered_customer_date - o.order_purchase_timestamp)) / 86400)::numeric, 1) AS avg_actual_delivery_days,
           ROUND(AVG(EXTRACT(EPOCH FROM (o.order_estimated_delivery_date - o.order_purchase_timestamp)) / 86400)::numeric, 1) AS avg_promised_delivery_days
    FROM orders o JOIN customers c ON c.customer_id = o.customer_id
    WHERE o.order_status = 'delivered' AND o.order_delivered_customer_date IS NOT NULL
    GROUP BY c.customer_state ORDER BY avg_actual_delivery_days DESC
''', engine)

,customer_state,avg_actual_delivery_days,avg_promised_delivery_days
0,RS,14.6,16.4
1,RJ,14.3,16.0
2,DF,14.3,16.1
3,BA,14.3,15.9
4,SP,14.2,16.0
5,CE,14.1,15.9
6,PA,14.0,15.8
7,GO,14.0,15.7
8,MG,13.9,15.7
9,SC,13.9,15.7


**Finding:** Actual delivery consistently beats the promised estimate by ~1.5-2 days across every state, which is good news operationally but is also a sign the estimated-delivery-date model is padded more than it needs to be — tightening the promise could be a low-risk way to improve the checkout experience without changing logistics.

## Conclusion

This notebook covers the full loop: pull → clean (with every drop/transform logged and justified) → relational
schema design → load into Postgres with real PK/FK constraints → 8 SQL business questions with findings written
for a business audience, not just a query result. See `README.md` for the question → query → finding table and
`queries.sql` for the standalone SQL.
